include/rosparam_shortcuts/parameter.h

```diff
@@ -0,0 +1,245 @@
+	// std::string<std::string, std::string> registered_parameter_callbacks_ids_;
+}
+
+struct ExampleConfig : public NodeConfig
+{
+	ExampleConfig(const rclcpp::Node::SharedPtr& node) : NodeConfig(node) {
+		// declare simple parameters just like that
+		declare(int1_param);
+		declare(int2_param);
+		declare(int3_param);
+
+		// "double_param", add some description and validity checks
+		declare(double_param)
+		    .describe("This is a very important double parameter", "should be greater than 0 and not 7")
+		    .warnIf([](const double& val, std::string& message) {
+			    message = "Value is 7, are you sure this is a good value?";
+			    return val == 7;
+		    })
+		    .errorIf([](const double& val, std::string& message) {
+			    message = "Value must not be negative!";
+			    return val < 0;
+		    });
+
+		// "string_param", use a parameter error condition "EMPTY"
+		declare(string_param)
+		    .describe("meh")
+		    .warnIf([](const std::string& val, std::string& message) {
+			    message = "string is much too long";
+			    return val.size() > 100;
+		    })
+		    .errorIf(EMPTY);
+
+		// "double_map_param", validate keys and values, reject if not valid
+		declare(double_map_param).describe("6-dof joint state").errorIf([](const auto& values, std::string& message) {
+			message = "Invalid joint state map values, expected 6-dof";
+			return values.size() != 6 || checkJointNamesInRobotModel(values);  // from somewhere
+		});
+
+		//
+		// can already check for errors/warnings here...
+		//
+	}
+
+	Parameter<double> double_param{ "double_param" };
+	Parameter<std::vector<std::string>> string_param{ "string_param" };
+	Parameter<std::map<std::string, double>> double_map_param{ "double_map_param" };
+	Parameter<int> int1_param{ "int1_param" };
+	Parameter<int> int2_param{ "int2_param" };
+	Parameter<int> int3_param{ "int3_param" };
+}
+
+void useConfig(const rclcpp::Node::SharedPtr& node)
+{
+	// use non-const config for initializing callbacks
+	// For safety, it should only be accessed as const in nodes
+	const ExampleConfig config(node);
+
+	// access copy of value
+	double value = config.double_param();
+
+	// ...
+
+	// React to parameter changes (sub is a bad example, but planning pipeline, plugin etc would work well)
+	auto sub = rclcpp::create_subscription(config.string_param, ...);  // sub managed somewhere else
+	config.string_param->onChanged(
+	    [&](const std::string& value) { sub = rclcpp::create_subscription(config.value, ...); });
+
+	// ...
+
+	auto robot_state = /* get robot state copy from somewhere else */;
+	auto update_robot_state = [this, robot_state](const std::map<std::string, double> values) {
+		for (const auto& [name, val] : values)
+			robot_state->setVariablePosition(name, val);
+
+		// other thread-safe member function somewhere
+		this->updateRobotState(robot_state);
+	};
+	config.double_map_param.apply(update_robot_state);
+	config.double_map_param.onChanged(update_robot_state);
```

Comment:
I like what this could enable.  ROS2 is supposed to be more event driven than ros1 and this will enable more event driven behaviors.  I also like that the values are passed in by value here so you don't have to worry about the issues around references and them being updated somewhere else at the same time.

Reply:
Right? ;) ... huh I guess I missed the reference, the `apply()` and `onChange()` calls would be read-locked so using const-ref should be fine. I mean it can work both ways, copy or const-ref.